### DISCUSSION TITLE
Swift vs ObjectiveC wrapper

### DISCUSSION THREAD

**Discussion by axxel** [zxing-cpp/zxing-cpp#726] [General]
Some random thought that just crossed my mind: Would a Swift wrapper (around the C-API) as a replacement for the current ObjectiveC wrapper be a net or even absolute win (apart from the time it takes to write it...). The idea is inspired by how 'clean' the C# wrapper based on the C-API is compared to the ObjectiveC wrapper based on the C++-API. The former is just [one file](https://github.com/zxing-cpp/zxing-cpp/blob/master/wrappers/dotnet/ZXingCpp/ZXingCpp.cs) with 350 lines of code. That includes about 60 lines of function declarations that Swift can apparently infer directly from a c-header file. Compare that to currently short of 1000 lines of ObjectiveC code.

Questions to the iOS crowd (@parallaxe, @markusfisch, @alexmanzer, @benjohnde, @ryantrem):
 * Is the use of a Swift module/class from ObjectiveC as seamless and convenient as a native ObjectiveC one?
 * Is support for application code in ObjectiveC even a thing anymore nowadays?

**Comment by markusfisch** [comment]
> Would a Swift wrapper (around the C-API) as a replacement for the current ObjectiveC wrapper be a net or even absolute win (apart from the time it takes to write it...)

At some point for sure, but it's still too early to switch to a Swift only wrapper as this would require Swift 5.9/Xcode15.

And from the consumer's point of view, it makes no difference whether the wrapper is written in Swift or Objective-C. So switching to Swift now would only restrict usage, but wouldn't have any advantage (apart from having a slightly smaller codebase).

> Is the use of a Swift module/class from ObjectiveC as seamless and convenient as a native ObjectiveC one?

More or less. You'd need some annotations, but this is not unusal in Objective-C. Here's the official documentation for using Swift in Objective-C to give you an idea: https://developer.apple.com/documentation/swift/importing-swift-into-objective-c

> Is support for application code in ObjectiveC even a thing anymore nowadays?

Depends on who you ask, of course 😉  While Swift is definitely the first choice, there are still a lot of Objective-C codebases out there. And it's not deprecated by Apple, so Objective-C is still a thing in general.

**Reply by axxel** [reply]
> At some point for sure, but it's still too early to switch to a Swift only wrapper as this would require Swift 5.9/Xcode15.

I don't understand: writing the Swift wrapper based on the automatic c-interop support requires that or being able to access the wrapper from ObjectiveC requires that? (I have not found anything supporting either of those theories.)

> And from the consumer's point of view, it makes no difference whether the wrapper is written in Swift or Objective-C.

Yes.

> So switching to Swift now would only restrict usage

In what sense?

> More or less. You'd need some annotations, but this is not unusal in Objective-C. Here's the official documentation for using Swift in Objective-C to give you an idea: https://developer.apple.com/documentation/swift/importing-swift-into-objective-c

It seems to work seamlessly based on that documentation. Maybe you'd need to do some forward declarations to break cyclic import dependencies (like in c++ in a similar situation, I guess).

> Depends on who you ask, of course 😉 While Swift is definitely the first choice, there are still a lot of Objective-C codebases out there. And it's not deprecated by Apple, so Objective-C is still a thing in general.

Is your company still maintaining ObjectiveC application code (that uses zxing-cpp)?

**Reply by markusfisch** [reply]
> I don't understand: writing the Swift wrapper based on the automatic c-interop support requires that or being able to access the wrapper from ObjectiveC requires that? (I have not found anything supporting either of those theories.)

Both - C++ interoperability is introduced in Version 5.9, the most current version of Swift. You can check it's current state and limitations [here](https://www.swift.org/documentation/cxx-interop/status/#constraints-and-limitations) and on [GitHub](https://github.com/apple/swift/issues/66159).

The very first thing you'll find there is this note:
> Note: Swift 5.9 is still undergoing development, and therefore some limitations and issues listed here will be fixed before Swift 5.9 is release. 

So this is what I'd call too early 😉 

> In what sense?

That it would require Swift 5.9 to build and to use the wrapper.

> Is your company still maintaining ObjectiveC application code (that uses zxing-cpp)?

Not at the moment, at least as far as I know, @alexmanzer please correct me if I'm wrong.

**Reply by axxel** [reply]
I see, good that I asked: there is a misunderstanding. I was talking about a wrapper around the C-API. I would think that support for direct C-interoperability was available from Swift 1.0.

**Reply by markusfisch** [reply]
Ah, sorry, I should've read more carefully 🙈 

That is an interesting question indeed! 🤔

**Reply by benjohnde** [reply]
That could work, haven't used the Swift C-interoperability layer yet and I don't know how it is actually consumed on the Swift-end. For a larger ML-module, we are still relying on a lot of ObjC glue code, generated via [djinni](https://github.com/cross-language-cpp/djinni-generator). It was not feasible for us to work around some limitations as we are in need of C++. We had some experiments with a C wrapper, but in the end the pain was not worth it. We will probably shift all to Swift soon. For a broader open source library, I would wait until ObjC is deprecated or at least two more years for a broader Swift adoption - pure gut feeling here.

For submitting apps you are mostly forced to use the latest Xcode but not the latest Swift.

The current state is working and I don't feel that a rewrite is beneficial but like the idea of reducing code for the sake of maintenance.

**Reply by alexmanzer** [reply]
> Not at the moment, at least as far as I know, @alexmanzer please correct me if I'm wrong.

No, productive projects that have been running in ObjC for years "out there" still use the zxing-objc variant or the native iOS code reader; we have not done anything in the direction of porting to zxing-cpp.
If we have to touch such legacy apps again, we would write them in Swift and then of course also use the native barcode reader from iOS or zxing-cpp for special cases (such as binary payloads).

Also in my opinion, ObjC is not dead yet. We only stumbled across it last week when we were looking into the iOS <-> .Net Maui interaction topic. Again, this requires a wrapper, which is provided in ObjC.

**Comment by axxel** [comment]
It has been 2 years since I was floating the idea of doing a Swift (over C-API) rewrite and with the 3.0 release and a now more outdated than ever ObjectiveC wrapper, I finally went ahead. That new wrapper exposes the full API, just as the Rust and .NET wrappers do.

The code can be found (and commented on) here: #1061 

Due to the fact that SwiftPM and/or CocoaPods wants to have one project at the top-level, it looks like I would have to do a "hard switch" from the old ObjectiveC based one to the new one with a completely new API. That sounds consumer unfriendly at first sight, but I'm not an iOS/macOS developer. The old wrapper code can still be part of the repo but I don't know how one would get access to it. Any ideas?

**Reply by axxel** [reply]
I think I have an idea on how to keep both the old and new Apple/iOS/etc. wrappers accessible. In the podspec file it will be via suspecs. This brings me to the question of why @benjohnde introduced a separate `Core` subspec in https://github.com/zxing-cpp/zxing-cpp/commit/69173797312e737e7dca04002d09940cf565dd84? To my understanding, this is just a C++ target, meaning it does not provide any usable code without providing one's own ObjectiveC++ wrapper. IIRC then @ryantrem mentioned at some point in time that he is using that?

**Reply by ryantrem** [reply]
It's been a couple years, but back then what we were doing was using the c++ API only because we were implementing a native BarcodeDetector being injected into a JavaScript runtime (React Native), so we didn't want any ObjC getting in the way of cross platform C++ code.

**Reply by axxel** [reply]
@ryantrem Is that still a thing for you today? I was not aware that one could/would use CocoaPods for seemingly non-Apple platform code? If I added a new subspec that contains the new libzint backend (and discards the old MultiFormatWriter one), would you still want that to be accessible as a plain c++ target without any Swift wrapper code?

**Reply by ryantrem** [reply]
I've been off the project for a couple years and different people have taken over, but I'm sure the desire would still be to have a pure c++ solution. It's the same c++ code, just being built on/for different platforms.

**Reply by axxel** [reply]
Could you point them to this discussion? I would appreciate any feedback on the new wrapper code and/or the question of how to provide both or a user friendly transition.

**Reply by axxel** [reply]
Update: the latest version of #1061 contains a podspec with 3 subspecs: `LegacyCore` and `LegacyWrapper` that should behave just like the old ones did and a new `Swift` one that corresponds to the `ZXingCpp` target in the Project.swift file. The latter does currently not provide access to the ObjectiveC wrapper.

**Comment by axxel** [comment]
@farfromrefug pointed out that he'd prefer continued access to an ObjectiveC consumable wrapper. He was also interred in getting access to the libzint backend from ObjectiveC. My current PR does not provide that.

I just learned about the requirements to make Swift code accessible from ObjectiveC code and the LLM of my choice suggested to provide a thin `@obj`-tified Swift wrapper around the native one to get ObjectiveC access. Whether that makes sense or whether it would be better to "update" the current native ObjectiveC code to allow for that is unclear to me atm. The latter would mean an API braking change.

**Reply by farfromrefug** [reply]
@axxel thanks a lot for taking my point into account! 
Yes @objc decorator is the best approach. The tricky part is that some things can't be "wrapped" with @objc. I remember some issues with swift enum or struct

**Comment by axxel** [comment]
A lengthy discussion with @jknlsn in #1061 about `Sendable` Option classes made me question my approach there in general. The `ReaderOptions` thing is only there because C++ does not support named, optional parameters and those options classes are a workaround for that. Python does support those and I never bothered introducing an `Option` class there. Swift supports them as well. My stated goal was to write each wrapper as close to as one would write a native library as possible. So why not simply support the following use cases:

```Swift
// writing
let image = try Barcode(text, format: .qrCode).toImage(scale: 4)

// reading
let barcodes = try readBarcodes(from: image, formats: [.qrCode, .pdf417], tryInvert: false)
```

Then there are no `Option` classes required at all and the question of how those parameters are collected and/or passed between execution contexts is completely outside the scope of the wrapper itself.

I have not evaluated how this would play out for the ObjectiveC layer mentioned above. But from my point of view, that one can be different if need be.

Since I have not heard anything from @alexmanzer, @benjohnde, @markusfisch regarding this new wrapper idea, forgive me for calling you out explicitly here. If you are still interested in the subject, please let me know what you think regarding this new Swift replacement  idea.

**Reply by axxel** [reply]
@jknlsn do you have an opinion regarding the free-function / Option-less approach outlined here?

**Reply by jknlsn** [reply]
Sorry @axxel, completely lost track of this. I'd be all for it personally, sounds good to me!

**Comment by markusfisch** [comment]
Sorry for the late reply, we’re so very busy at the moment.

In general we would certainly welcome no longer having to deal with the Objective-C mapping code. A Swift interface may be more of an issue for React Native or .NET MAUI, where Objective-C is still very much used as lowest common denominator, but we’re using the library in native code only.

**Reply by axxel** [reply]
Not quite sure I understand: you would be interested in switching to a new Swift wrapper or you would not because you technically are still depending on it? If it is the latter, would you switch if there was this mentioned ObjectiveC compatibility layer above the "swifty" Swift layer?

**Reply by markusfisch** [reply]
We would definitely be interested in switching to a new Swift wrapper! Sorry for being unclear 🙈

I just wanted to mention that Swift can still be an issue for cross-plattform solutions like React Native or .NET MAUI (we’ve been there with other libraries, unfortunately).

**Reply by farfromrefug** [reply]
That s actually why i want @objc in swift or objective-c wrapper. I am one of the maintainers of Nativescript and our runtime still need objc
